In [1]:
import pandas as pd
import numpy as np

In [2]:
from sklearn.metrics import classification_report
from sklearn.metrics import f1_score


In [3]:
df_staff = pd.read_csv('cj_empN.csv')
df_nonstaff = pd.read_csv('cj_empY.csv')
df = pd.read_csv('cj.csv')

In [16]:
df

scd                      product_name  net_order_qty  \
0      20230124153976                  잔칫집 식혜 240ml 30입              1   
1      20230124155563                 백설 한입쏙 비엔나 120g*2              1   
2      20230125158386                    비비고 왕교자 1.05kg              1   
3      20230126164638                 고메 바삭쫄깃한 탕수육 900g              1   
4      20230125159705                      햇반 매일잡곡밥210g             20   
...               ...                               ...            ...   
43715  20230102972321                고메 거멍 모짜체다핫도그 340g              1   
43716  20230102972720         [앱전용특가]비비고 차돌된장찌개 460gX4개              1   
43717  20230101964953         [앱전용특가]비비고 차돌된장찌개 460gX4개              1   
43718  20230101970142         [식물성]고메 플랜테이블 함박스테이크 150g              1   
43719  20230101966180  리턴업 전립소 쏘팔메토 골드 1000mgX60캡슐(2개월)              1   

       net_order_amt  gender  age_grp employee_yn  prime_yn  weekday  \
0           9.803170       0        2           Y         0        1   
1           8.256607       1        3           Y         0        1   
2           9.348449       0        4           N         0        2   
3           9.667259       0        4           N         1        3   
4           9.994653       1        4           N         1        2   
...              ...     ...      ...         ...       ...      ...   
43715       8.648397       1        5           Y         1        0   
43716       9.639327       0        3           N         0        0   
43717       9.639327       0        4           N         0        6   
43718       7.939872       1        4           Y         1        6   
43719      10.406140       1        5           N         1        6   

       weekday2  event_product  row_num  category  newyear  category_count  \
0             1              0        1         1        0               4   
1             1              0        1         1        0               5   
2             1              0        1         3        0               1   
3             1              0        1         3        0               3   
4             1              0        1         6        0               3   
...         ...            ...      ...       ...      ...             ...   
43715         1              0        1         1        0               4   
43716         1              1        1        10        0               2   
43717         0              1        1        10        0               1   
43718         0              0        1         5        0               3   
43719         0              0        1         8        0               2   

       most_product  
0                 0  
1                 0  
2                 1  
3                 1  
4                 2  
...             ...  
43715             0  
43716             3  
43717             3  
43718             1  
43719             3  

[43720 rows x 16 columns]

In [57]:
df_nonstaff.drop('scd',axis=1,inplace=True)
df_staff.drop('scd',axis=1,inplace=True)

## 로지스틱 회귀

- 일반회원

In [58]:
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

In [59]:
df_nonstaff.columns

Index(['gender', 'age_grp', 'order_date', 'total_qty', 'total_amt',
       'event_product', 'amt/qty', 'prime_yn', 'ctg_0', 'ctg_1', 'ctg_2',
       'ctg_3'],
      dtype='object')

In [113]:
X = df_nonstaff[['gender', 'age_grp', 'order_date', 'total_qty', 'total_amt',
       'event_product', 'amt/qty', 'ctg_0', 'ctg_1', 'ctg_2',
       'ctg_3']]
y = df_nonstaff['prime_yn']

In [114]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=111)

In [115]:
# 로지스틱 회귀 모델 생성 및 훈련
model = LogisticRegression()
model.fit(X_train, y_train)

# 모델 예측
y_pred = model.predict(X_test)

In [65]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [66]:
y_pred
# 왜 다 1로 하지...

array([0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1,
       0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1,
       0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1,
       1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,

In [116]:
cfreport = classification_report(y_test, y_pred)
print("분류예측 레포트:\n", cfreport)

f1 = f1_score(y_test, y_pred)
print("f1 score:\n", f1)

분류예측 레포트:
               precision    recall  f1-score   support

           0       0.52      0.24      0.33       357
           1       0.65      0.87      0.74       586

    accuracy                           0.63       943
   macro avg       0.58      0.55      0.53       943
weighted avg       0.60      0.63      0.59       943

f1 score:
 0.7428571428571429


In [117]:
# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

Accuracy: 0.6277836691410392
Recall: 0.8651877133105802
Precision: 0.6508344030808729
F1 Score: 0.7428571428571429


- 임직원

In [68]:
X = df_staff.drop('prime_yn',axis=1)
y = df_staff['prime_yn']

In [69]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=100)

In [70]:
# 로지스틱 회귀 모델 생성 및 훈련
model = LogisticRegression()
model.fit(X_train, y_train)

# 모델 예측
y_pred = model.predict(X_test)

In [71]:
# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

Accuracy: 0.617003367003367
Recall: 0.3273092369477912
Precision: 0.5759717314487632
F1 Score: 0.41741357234314985


## 랜덤포레스트

- 일반회원

In [72]:
from sklearn.ensemble import RandomForestClassifier

In [73]:
from sklearn.ensemble import RandomForestClassifier

# Random Forest
rf = RandomForestClassifier()
rf.fit(X_train, y_train)

# 테스트 데이터로 예측 수행
pred = rf.predict(X_test)

# 정확도 평가
accuracy = accuracy_score(y_test, pred)
print("Accuracy:", accuracy)

Accuracy: 0.6127946127946128


In [74]:
# 정확도 계산
accuracy = accuracy_score(y_test, pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, pred)
print("F1 Score:", f1)

Accuracy: 0.6127946127946128
Recall: 0.4357429718875502
Precision: 0.547979797979798
F1 Score: 0.4854586129753915


- 임직원

In [118]:
X = df_staff.drop('prime_yn',axis=1)
y = df_staff['prime_yn']

In [119]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=100)

In [120]:
# 랜덤 포레스트 모델 생성
model = RandomForestClassifier()

# 파라미터 그리드 설정
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 5, 10],
    'min_samples_split': [2, 5, 10],
}

# 그리드 서치를 사용한 파라미터 조정
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='f1')
grid_search.fit(X_train, y_train)

# 최적의 파라미터와 그에 대한 모델의 성능 출력
print("Best Parameters:", grid_search.best_params_)
print("Best Accuracy:", grid_search.best_score_)

# 검증 데이터에 대한 예측
y_pred = best_model.predict(X_test)

Best Parameters: {'max_depth': None, 'min_samples_split': 5, 'n_estimators': 300}
Best Accuracy: 0.5078819465216403


In [121]:
cfreport = classification_report(y_test, y_pred)
print("분류예측 레포트:\n", cfreport)

f1 = f1_score(y_test, y_pred)
print("f1 score:\n", f1)

분류예측 레포트:
               precision    recall  f1-score   support

           0       0.66      0.20      0.31       690
           1       0.44      0.86      0.58       498

    accuracy                           0.48      1188
   macro avg       0.55      0.53      0.44      1188
weighted avg       0.57      0.48      0.42      1188

f1 score:
 0.5789830508474576


In [122]:
# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

Accuracy: 0.4772727272727273
Recall: 0.857429718875502
Precision: 0.43705220061412486
F1 Score: 0.5789830508474576


## 나이브 베이즈

- 일반회원

In [87]:
Xn = df_nonstaff.drop('prime_yn',axis=1)
yn = df_nonstaff['prime_yn']

In [88]:
X_train, X_test, y_train, y_test = train_test_split(Xn, yn, test_size=0.2, random_state=111)

In [89]:
from sklearn.naive_bayes import GaussianNB

In [90]:
model = GaussianNB()
model.fit(X_train, y_train)

# 테스트 데이터에 대한 예측
y_pred = model.predict(X_test)

In [91]:
y_pred

array([0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1,
       0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1,
       1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0,
       0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1,
       1, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1,
       1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 1,
       1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0,

In [92]:
# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

Accuracy: 0.5323435843054083
Recall: 0.4948805460750853
Precision: 0.6666666666666666
F1 Score: 0.5680705190989227


## XGBoost

- 일반회원

In [99]:
import xgboost as xgb

In [94]:
# XGBoost
xgb = xgb.XGBClassifier()
xgb.fit(X_train, y_train)

# 테스트 데이터로 예측 수행
y_pred = xgb.predict(X_test)

In [95]:
# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

Accuracy: 0.6214209968186638
Recall: 0.7559726962457338
Precision: 0.6742770167427702
F1 Score: 0.7127916331456154


- 임직원

In [96]:
Xe = df_staff.drop('prime_yn',axis=1)
ye = df_staff['prime_yn']

In [97]:
X_train, X_test, y_train, y_test = train_test_split(Xe, ye, test_size=0.2, random_state=100)

In [100]:
# XGBoost
xgb = xgb.XGBClassifier(random_state=111)
xgb.fit(X_train, y_train)

# 테스트 데이터로 예측 수행
y_pred = xgb.predict(X_test)

In [101]:
# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

Accuracy: 0.6481481481481481
Recall: 0.46586345381526106
Precision: 0.6041666666666666
F1 Score: 0.526077097505669


## XGBoost 파라미터 최적화

- 일반회원

In [102]:
Xn = df_nonstaff.drop('prime_yn',axis=1)
yn = df_nonstaff['prime_yn']

In [103]:
X_train, X_test, y_train, y_test = train_test_split(Xn, yn, test_size=0.2, random_state=100)

In [104]:
from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV, train_test_split

In [105]:
# XGBoost 모델 생성
model = XGBClassifier()

# 파라미터 그리드 설정
param_grid = {
    'max_depth': [3, 5, 7],
    'learning_rate': [0.1, 0.01, 0.001],
    'n_estimators': [100, 200, 300],
}

# 그리드 서치를 사용한 파라미터 조정
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='f1')
grid_search.fit(X_train, y_train)

# 최적의 파라미터와 그에 대한 모델의 성능 출력
print("Best Parameters:", grid_search.best_params_)
print("Best Score:", grid_search.best_score_)

# 최적의 파라미터로 모델 재학습
best_model = grid_search.best_estimator_
best_model.fit(X_train, y_train)

# 검증 데이터에 대한 예측
y_pred = best_model.predict(X_test)

Best Parameters: {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 300}
Best Score: 0.7448673929806707


In [106]:
# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

Accuracy: 0.6033934252386002
Recall: 0.9446494464944649
Precision: 0.5981308411214953
F1 Score: 0.7324749642346209


## 가장 좋은 값
- 표준화
- XGBoost for nonstaff

In [7]:
df_staff

scd  gender  age_grp  order_date  total_qty   total_amt  \
0     20230101963244       0        4           0          3   10.234373   
1     20230101963277       1        3           0          1    9.190036   
2     20230101963302       0        4           0         17   83.446280   
3     20230101963319       1        4           0         23   73.452653   
4     20230101963336       0        3           0         30  119.943496   
...              ...     ...      ...         ...        ...         ...   
5935  20230131216709       0        4           1          6   27.526347   
5936  20230131216756       0        4           1          3   18.314758   
5937  20230131216771       0        4           1          2    9.125762   
5938  20230131216842       1        3           1          1   10.176411   
5939  20230131216844       0        4           1          8   44.927942   

      event_product    amt/qty  prime_yn  ctg_0  ctg_1  ctg_2  ctg_3  
0                 0   3.411458         1      0      1      0      0  
1                 0   9.190036         1      0      0      1      0  
2                 0   7.574948         0      1      1      1      1  
3                 0   4.093077         0      1      1      1      0  
4                 0   6.208069         1      1      1      1      1  
...             ...        ...       ...    ...    ...    ...    ...  
5935              0   7.037744         1      1      0      0      1  
5936              0   6.796144         0      0      1      0      0  
5937              0   4.562881         1      1      0      0      0  
5938              1  10.176411         0      0      0      1      0  
5939              0   6.307187         0      0      1      0      1  

[5940 rows x 13 columns]

In [6]:
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler

In [8]:
X = df_staff[['total_amt','amt/qty','total_qty']]

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [9]:
X_scaled

array([[-0.80552985, -1.6850602 , -0.45570644],
       [-0.83398621,  0.6800985 , -0.66674579],
       [ 1.18936654,  0.01904701,  1.02156902],
       ...,
       [-0.83573754, -1.21378517, -0.56122611],
       [-0.8071092 ,  1.08381978, -0.66674579],
       [ 0.1398092 , -0.49984485,  0.07189194]])

In [10]:
df_staff[['total_amt','amt/qty','total_qty']] = X_scaled

In [11]:
df_staff

scd  gender  age_grp  order_date  total_qty  total_amt  \
0     20230101963244       0        4           0  -0.455706  -0.805530   
1     20230101963277       1        3           0  -0.666746  -0.833986   
2     20230101963302       0        4           0   1.021569   1.189367   
3     20230101963319       1        4           0   1.654687   0.917058   
4     20230101963336       0        3           0   2.393325   2.183852   
...              ...     ...      ...         ...        ...        ...   
5935  20230131216709       0        4           1  -0.139147  -0.334354   
5936  20230131216756       0        4           1  -0.455706  -0.585353   
5937  20230131216771       0        4           1  -0.561226  -0.835738   
5938  20230131216842       1        3           1  -0.666746  -0.807109   
5939  20230131216844       0        4           1   0.071892   0.139809   

      event_product   amt/qty  prime_yn  ctg_0  ctg_1  ctg_2  ctg_3  
0                 0 -1.685060         1      0      1      0      0  
1                 0  0.680098         1      0      0      1      0  
2                 0  0.019047         0      1      1      1      1  
3                 0 -1.406075         0      1      1      1      0  
4                 0 -0.540414         1      1      1      1      1  
...             ...       ...       ...    ...    ...    ...    ...  
5935              0 -0.200830         1      1      0      0      1  
5936              0 -0.299716         0      0      1      0      0  
5937              0 -1.213785         1      1      0      0      0  
5938              1  1.083820         0      0      0      1      0  
5939              0 -0.499845         0      0      1      0      1  

[5940 rows x 13 columns]

In [35]:
Xn = df_staff.drop(['prime_yn','event_product'],axis=1)
yn = df_staff['prime_yn']

In [36]:
X_train, X_test, y_train, y_test = train_test_split(Xn, yn, test_size=0.2, random_state=100)

In [37]:
from xgboost import XGBClassifier
from sklearn.model_selection import GridSearchCV, train_test_split

In [38]:
# XGBoost 모델 생성
model = XGBClassifier()

# 파라미터 그리드 설정
param_grid = {
    'max_depth': [3, 5, 7],
    'learning_rate': [0.1, 0.01, 0.001],
    'n_estimators': [100, 200, 300],
}

# 그리드 서치를 사용한 파라미터 조정
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='f1')
grid_search.fit(X_train, y_train)

# 최적의 파라미터와 그에 대한 모델의 성능 출력
print("Best Parameters:", grid_search.best_params_)
print("Best Score:", grid_search.best_score_)

# 최적의 파라미터로 모델 재학습
best_model = grid_search.best_estimator_
best_model.fit(X_train, y_train)

# 검증 데이터에 대한 예측
y_pred = best_model.predict(X_test)

Best Parameters: {'learning_rate': 0.1, 'max_depth': 7, 'n_estimators': 300}
Best Score: 0.5220479595525565


In [39]:
cfreport = classification_report(y_test, y_pred)
print("분류예측 레포트:\n", cfreport)

f1 = f1_score(y_test, y_pred)
print("f1 score:\n", f1)

분류예측 레포트:
               precision    recall  f1-score   support

           0       0.65      0.76      0.70       690
           1       0.57      0.43      0.49       498

    accuracy                           0.62      1188
   macro avg       0.61      0.60      0.60      1188
weighted avg       0.62      0.62      0.61      1188

f1 score:
 0.4920273348519362


In [40]:
# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

Accuracy: 0.6245791245791246
Recall: 0.43373493975903615
Precision: 0.5684210526315789
F1 Score: 0.4920273348519362


- 임직원은 다른 모델 사용해야할 것
- 랜덤포레스트가 더 좋다?? -> 파라미터 조정 후 0.58 나옴
- event_product 없애니 더 성능 안좋아짐

In [54]:
Xn = df_staff.drop('prime_yn',axis=1)
yn = df_staff['prime_yn']

In [55]:
X_train, X_test, y_train, y_test = train_test_split(Xn, yn, test_size=0.2, random_state=100)

In [56]:
# 랜덤 포레스트 모델 생성
model = RandomForestClassifier()

# 파라미터 그리드 설정
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 5, 10],
    'min_samples_split': [2, 5, 10],
}

# 그리드 서치를 사용한 파라미터 조정
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='f1')
grid_search.fit(X_train, y_train)

# 최적의 파라미터와 그에 대한 모델의 성능 출력
print("Best Parameters:", grid_search.best_params_)
print("Best Accuracy:", grid_search.best_score_)

# 검증 데이터에 대한 예측
y_pred = best_model.predict(X_test)

Best Parameters: {'max_depth': None, 'min_samples_split': 5, 'n_estimators': 100}
Best Accuracy: 0.4974749819731305


ValueError: Feature shape mismatch, expected: 11, got 12

In [ ]:
cfreport = classification_report(y_test, y_pred)
print("분류예측 레포트:\n", cfreport)

f1 = f1_score(y_test, y_pred)
print("f1 score:\n", f1)

In [ ]:
# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

In [78]:
df

scd                      product_name  net_order_qty  \
0      20230124153976                  잔칫집 식혜 240ml 30입              1   
1      20230124155563                 백설 한입쏙 비엔나 120g*2              1   
2      20230125158386                    비비고 왕교자 1.05kg              1   
3      20230126164638                 고메 바삭쫄깃한 탕수육 900g              1   
4      20230125159705                      햇반 매일잡곡밥210g             20   
...               ...                               ...            ...   
43715  20230102972321                고메 거멍 모짜체다핫도그 340g              1   
43716  20230102972720         [앱전용특가]비비고 차돌된장찌개 460gX4개              1   
43717  20230101964953         [앱전용특가]비비고 차돌된장찌개 460gX4개              1   
43718  20230101970142         [식물성]고메 플랜테이블 함박스테이크 150g              1   
43719  20230101966180  리턴업 전립소 쏘팔메토 골드 1000mgX60캡슐(2개월)              1   

       net_order_amt  gender  age_grp employee_yn  prime_yn  weekday  \
0           9.803170       0        2           Y         0        1   
1           8.256607       1        3           Y         0        1   
2           9.348449       0        4           N         0        2   
3           9.667259       0        4           N         1        3   
4           9.994653       1        4           N         1        2   
...              ...     ...      ...         ...       ...      ...   
43715       8.648397       1        5           Y         1        0   
43716       9.639327       0        3           N         0        0   
43717       9.639327       0        4           N         0        6   
43718       7.939872       1        4           Y         1        6   
43719      10.406140       1        5           N         1        6   

       weekday2  event_product  row_num  category  newyear  category_count  \
0             1              0        1         1        0               4   
1             1              0        1         1        0               5   
2             1              0        1         3        0               1   
3             1              0        1         3        0               3   
4             1              0        1         6        0               3   
...         ...            ...      ...       ...      ...             ...   
43715         1              0        1         1        0               4   
43716         1              1        1        10        0               2   
43717         0              1        1        10        0               1   
43718         0              0        1         5        0               3   
43719         0              0        1         8        0               2   

       most_product  
0                 0  
1                 0  
2                 1  
3                 1  
4                 2  
...             ...  
43715             0  
43716             3  
43717             3  
43718             1  
43719             3  

[43720 rows x 16 columns]

In [18]:
X = df_nonstaff[['total_amt','amt/qty','total_qty']]

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [19]:
X_scaled

array([[-0.00857481,  0.04415783, -0.15434383],
       [-0.44211005,  0.74288605, -0.54585216],
       [ 0.12198538,  0.20445538, -0.15434383],
       ...,
       [-0.0539896 ,  0.64271469, -0.35009799],
       [-0.22177682,  0.81811506, -0.44797508],
       [ 0.07578048,  0.41418117, -0.25222091]])

In [20]:
df_nonstaff[['total_amt','amt/qty','total_qty']] = X_scaled

In [107]:
df_nonstaff

gender  age_grp  order_date  total_qty  total_amt  event_product  \
0          1        3           0  -0.154344  -0.008575              2   
1          0        4           0  -0.545852  -0.442110              2   
2          1        3           0  -0.154344   0.121985              0   
3          0        3           0  -0.545852  -0.622162              0   
4          1        2           0  -0.154344   0.154134              0   
...      ...      ...         ...        ...        ...            ...   
4708       0        5           1  -0.154344  -0.261427              1   
4709       1        3           1   0.432919   0.904903              6   
4710       1        4           1  -0.350098  -0.053990              0   
4711       1        5           1  -0.447975  -0.221777              3   
4712       1        5           1  -0.252221   0.075780              4   

       amt/qty  prime_yn  ctg_0  ctg_1  ctg_2  ctg_3  
0     0.044158         0      1      0      1      0  
1     0.742886         1      1      1      0      0  
2     0.204455         1      1      1      0      1  
3    -0.716926         1      0      0      0      1  
4     0.335082         1      1      1      0      1  
...        ...       ...    ...    ...    ...    ...  
4708 -0.608848         1      1      0      0      0  
4709  0.299230         1      1      1      0      1  
4710  0.642715         0      0      1      0      0  
4711  0.818115         1      1      0      1      0  
4712  0.414181         1      1      1      1      1  

[4713 rows x 12 columns]

In [108]:
Xn = df_nonstaff.drop('prime_yn',axis=1)
yn = df_nonstaff['prime_yn']

In [109]:
X_train, X_test, y_train, y_test = train_test_split(Xn, yn, test_size=0.2, random_state=100)

In [110]:
# XGBoost 모델 생성
model = XGBClassifier()

# 파라미터 그리드 설정
param_grid = {
    'max_depth': [3, 5, 7],
    'learning_rate': [0.1, 0.01, 0.001],
    'n_estimators': [100, 200, 300],
}

# 그리드 서치를 사용한 파라미터 조정
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=3, scoring='f1')
grid_search.fit(X_train, y_train)

# 최적의 파라미터와 그에 대한 모델의 성능 출력
print("Best Parameters:", grid_search.best_params_)
print("Best Score:", grid_search.best_score_)

# 최적의 파라미터로 모델 재학습
best_model = grid_search.best_estimator_
best_model.fit(X_train, y_train)

# 검증 데이터에 대한 예측
y_pred = best_model.predict(X_test)

Best Parameters: {'learning_rate': 0.01, 'max_depth': 3, 'n_estimators': 300}
Best Score: 0.7448673929806707


In [111]:
cfreport = classification_report(y_test, y_pred)
print("분류예측 레포트:\n", cfreport)

f1 = f1_score(y_test, y_pred)
print("f1 score:\n", f1)

분류예측 레포트:
               precision    recall  f1-score   support

           0       0.66      0.14      0.23       401
           1       0.60      0.94      0.73       542

    accuracy                           0.60       943
   macro avg       0.63      0.54      0.48       943
weighted avg       0.62      0.60      0.52       943

f1 score:
 0.7324749642346209


In [112]:
# 정확도 계산
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

# 재현율 계산
recall = recall_score(y_test, y_pred)
print("Recall:", recall)

# 정밀도 계산
precision = precision_score(y_test, y_pred)
print("Precision:", precision)

# F1 스코어 계산
f1 = f1_score(y_test, y_pred)
print("F1 Score:", f1)

Accuracy: 0.6033934252386002
Recall: 0.9446494464944649
Precision: 0.5981308411214953
F1 Score: 0.7324749642346209


- 임직원은 랜덤포레스트가 가장 잘나옴... 0.57
- 일반회원은 XGB가 0.73, 로지스틱 회귀가 0.74나옴..